tap_s3_csv/s3.py

```diff
@@ -171,7 +172,7 @@ def sample_files(config: Dict, table_spec: Dict, s3_files: Generator,
     :returns: Generator containing all samples as dicts
     """
     LOGGER.info("Sampling files (max files: %s)", max_files)
-    for s3_file in itertools.islice(s3_files, max_files):
+    for s3_file in more_itertools.tail(max_files, s3_files):
```

Comment:
How to values of the `s3_files` generator ordered?

I assume we'll take the last N files ordered alphabetically. That's not necessarily in sync with the files which are actually about to load. The files that is selected to load is using a different logic by checking the `LastModified` tag on the file. Code [here](https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/b1156c62f983afe4b44cabbd94a5e5eb258b5b4e/tap_s3_csv/s3.py#L210).
In some cases it's possible that we analyse different files that we are loading. 

Can it cause any problem?

Reply:
ignore this, I was wrong. `s3_files` is actually the files that is selected to load and it's not relevant

Reply:
edge case:

let's say 10 files are selected to load. Which ones we want to analyse? If we get only the last file then the structure may not be the same across every file that we want to load.

To solve it in a fully automated way I think we need sample all the files that is selected to load. And we should come across an algorithm to merge the different structures into one that fits for all of them.

Maybe it's not a relevant problem for now and unlikely to happen but worth to mention.

Reply:
@koszti do you think it's necessary to do this, then?

https://github.com/transferwise/pipelinewise-tap-s3-csv/pull/62/files#diff-5a76c66a3c73e1e27a33a762f66cd34154903d76839531d65056e9f2d1f43bf2R209

Reply:
On problem we have is the sampling doesn't take `start_date` into consideration, the function call here: https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/b1156c62f983afe4b44cabbd94a5e5eb258b5b4e/tap_s3_csv/s3.py#L82 is not providing the argument `modified_since`.

Reply:
Ah sorry I wrote my reply to your previous message before I reloaded and saw your newer comment 😅 
So, I agree that this doesn't solve the edge case and I'm not sure I quite understand data warehousing well enough to understand how schema changes over time are actually supposed to be handled... I'd be happy to collab on a more elegant solution in future though!

For now I'm definitely more inclined towards a quick fix, and I think generally a "naive sample" of more _recent_ files is better than a naive sample of the oldest files - because I believe schema changes are more likely to be **additive** rather than **reductive** in tabular data like CSVs.

However, even with this "quick fix" solution, do you think we need to actively sort the files by LastModified (as in my most recent commit), or is tailing rather than slicing enough (thus, relying on filenames)?

Are you concerned with the in-memory sorting in my most recent commit? I note the comment in the sync part of the codebase here which already comments on this and seems to make peace with it:
https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/master/tap_s3_csv/sync.py#L42

Reply:
Also, I think that picking the last X files is as problematic as picking the first X ones, we could instead get a random sample of size X, or have a X/2 from start and X/2 from end.

Reply:
picking random files is exactly as problematic as picking the first or last X files :)

Reply:
And increasing the likelihood is what we want. We can't scan thru all the files during discovery, by having a mixed samples of files: old, new and in between, you can get a closer idea of the file structure, would you miss some column X that appears once in a file Y among throusands? yes, but the responsibility to keep the files structure consistent lies on the data owner.

As for detecting schema changes, that would be during the sync, which this PR isn't touching.
, we have

Reply:
> On problem we have is the sampling doesn't take `start_date` into consideration, the function call here:
> 
> https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/b1156c62f983afe4b44cabbd94a5e5eb258b5b4e/tap_s3_csv/s3.py#L82
> 
> is not providing the argument `modified_since`.

I do agree that implementing support for `modified_since` in the sampling code path in `s3.py` is a more elegant solution; does my most recent commit look right?

To be clear, none of this has been properly tested yet, I'm not sure how to actually test locally 😅 

On the point of first vs. last vs. random - I don't think any of them are ideal, as per Peter's comment earlier I generally agree that a future version of this ought to sample all of them and identify structures to load all of the data correctly.
However, I don't think now is the time to do that; and I really do feel CSVs are far more likely to have additional columns _added_ over time than removed/modified - so could we go ahead with this for now, since it'll solve the current problem at least? 😄

Reply:
Ok, I think my latest commit should be better now - I've implemented the pass through of `modified_since` from the config (I was totally overthinking it, sorry!).

I've also now successfully tested this branch locally, so I can say it definitely works to sample only files newer than the `start_date` value, and also sorts them to sample the most recent of that set rather than the oldest of that set.

<img width="1327" alt="pipelinewise-tap-s3-csv 2021-07-14 17-30-31" src="https://user-images.githubusercontent.com/757713/125658752-a4f0571b-de1b-48c9-91a1-4fd2fa2b5714.png">

I believe this should now be good to go, so I'm resolving this thread; feel free to open another if you still think there's anything wrong with this as-is 😄